File: Pre_requisites/Patterns/cpp/pattern_9.cpp
/*

Print a diamond for given N as shown:

N = 3

Output:

  *
 ***
*****
*****
 ***
  *
  
*/

void nStarDiamond(int n) {
    int i,j,k,n_spaces,n_stars;

    for(i=1;i<n+1;i++){

        n_stars = 2*i-1;
        n_spaces = n-i;

        for(j=1;j<=n_spaces;j++){
            cout<<" ";
        }

        for(k=1;k<=n_stars;k++){
            cout<<"*";
        }
        cout<<endl;
    
    }

        for(i=n;i>0;i--){

        n_stars = 2*i-1;
        n_spaces = n-i;

        for(j=1;j<=n_spaces;j++){
            cout<<" ";
        }

        for(k=1;k<=n_stars;k++){
            cout<<"*";
        }
        cout<<endl;
    
    }
}

File: Pre_requisites/Patterns/python/pattern_9.py
'''
Problem statement:
Ninja was very fond of patterns. For a given integer ‘N’, he wants to make the Reverse N-Star Triangle.

Example:
Input: ‘N’ = 3

Output: 

  *
 ***
*****
*****
 ***
  *

'''

def nStarDiamond(n: int) -> None:

    for i in range(1,n+1):

        n_stars=2*i-1
        n_spaces=n-i

        print(" "*n_spaces+"*"*n_stars)
    
    for i in range(n,0,-1):

        n_stars=2*i-1
        n_spaces=n-i

        print(" "*n_spaces+"*"*n_stars)


#Single loop approach:

def nStarDiamond(n: int) -> None:

    for i in range(1,2*n+1):

        if i<n:
            n_stars=2*i-1
            n_spaces=n-i

        elif i==n:
            n_stars=2*i-1
            n_spaces=n-i
            print(" "*n_spaces+"*"*n_stars)

        else:
            n_stars=2*(2*n-i)-1
            n_spaces=i-n

        print(" "*n_spaces+"*"*n_stars)